File: 백준/Gold/2263. 트리의 순회/트리의 순회.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(1000003)

n = int(input())
inorder = list(map(int,input().split()))
postorder = list(map(int,input().split()))

class trees:
    def __init__(self):
        self.root=None
class node:
    def __init__(self,data):
        self.parent=None
        self.data = data
        self.left=None
        self.right=None


where = dict()
for i in range(n):
    where[inorder[i]]=i

def search(s,e):
    data = postorder.pop()
    # print(s,e,data)
    # print(postorder)
    if s==e:
        return node(data)
    else:
        nd = node(data)
        index = where[data]
        # print('index',index)
        if index+1<=e:
            nd.right = search(index+1,e)
        if index -1 >=s:
            nd.left= search(s,index-1)
        return nd
        
        
tree = trees()
tree.root = search(0,n-1)

def preorder(x:node):
    print(x.data, end=' ')
    if x.left:
        preorder(x.left)
    if x.right:
        preorder(x.right)

preorder(tree.root)

# (s,e)
# index,x
# x.right = (index+1,e)
# x.left = (s,index-1)

File: 백준/Gold/2263. 트리의 순회/트리의 순회.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] inorder, postorder;
    static Map<Integer, Integer> where = new HashMap<>();
    static int postIdx;
    static StringBuilder sb = new StringBuilder();

    static class Node {
        int data;
        Node left, right;

        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        inorder = new int[n];
        postorder = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            inorder[i] = Integer.parseInt(st.nextToken());
            where.put(inorder[i], i);
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            postorder[i] = Integer.parseInt(st.nextToken());
        }

        postIdx = n - 1;
        Node root = search(0, n - 1);
        preorder(root);
        System.out.println(sb.toString());
    }

    static Node search(int s, int e) {
        if (s > e) return null;

        int data = postorder[postIdx--];
        Node node = new Node(data);

        int index = where.get(data);

        node.right = search(index + 1, e);
        node.left = search(s, index - 1);

        return node;
    }

    static void preorder(Node node) {
        if (node == null) return;
        sb.append(node.data).append(" ");
        preorder(node.left);
        preorder(node.right);
    }
}